### PR TITLE
[2.1] use `random_int()` instead of `mt_rand()` to generate cryptographically secure pseudo-random integers

### PR DESCRIPTION
**Pull request by yyxx9988**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | no
| Breaks BC?    | yes
| Tests pass?   | yes
| Fixed issues  | no

like `random_bytes()`, PHP 7.0 introduces `random_int()` as a core function, consider migrating.

refs:
- offical site: https://secure.php.net/manual/en/function.random-int.php
- a userland implementation for php5.2 to 5.6 https://github.com/paragonie/random_compat
- benchmark (random_int vs mt_rand) https://gist.github.com/masakielastic/0f71d119911225b708bc

### CODE REVIEW

**Comment by FabrizioCaldarelli** [comment]
Why don't create a method  (for example, randomInt) in yii\base\Security that wraps both implementations ( below php 7 and over php 7 versions)? So this method could be used in other part of framework and applications.

**Comment by Renkas** [comment]
@FabrizioCaldarelli Yii 2.1 has minimum requirenment of php 7.1 so there is no need for that.

**Comment by FabrizioCaldarelli** [comment]
It could be useful also to prevent other future changes.

**Review by samdark — APPROVED** [review]

**Review by dynasource — APPROVED** [review]

File: framework/caching/DbCache.php [hunk 1]
```diff
@@ -256,7 +256,7 @@ protected function deleteValue($key)
      */
     public function gc($force = false)
     {
-        if ($force || mt_rand(0, 1000000) < $this->gcProbability) {
+        if ($force || random_int(0, 1000000) < $this->gcProbability) {
```
Comment (klimov-paul) [inline framework/caching/DbCache.php]: I am unsure it make sense to change `gc`. It has nothing to do with security or cryptographics.
Replacing `mt_rand` with `random_int` will degrade performance without purpose here.

File: framework/caching/FileCache.php [hunk 2]
```diff
@@ -232,7 +232,7 @@ protected function flushValues()
      */
     public function gc($force = false, $expiredOnly = true)
     {
-        if ($force || mt_rand(0, 1000000) < $this->gcProbability) {
+        if ($force || random_int(0, 1000000) < $this->gcProbability) {
```
Comment (klimov-paul) [inline framework/caching/FileCache.php]: Same here.

File: framework/mail/BaseMailer.php [hunk 3]
```diff
@@ -249,7 +249,7 @@ public function generateMessageFileName()
     {
         $time = microtime(true);
 
-        return date('Ymd-His-', $time) . sprintf('%04d', (int) (($time - (int) $time) * 10000)) . '-' . sprintf('%04d', mt_rand(0, 10000)) . '.eml';
+        return date('Ymd-His-', $time) . sprintf('%04d', (int) (($time - (int) $time) * 10000)) . '-' . sprintf('%04d', random_int(0, 10000)) . '.eml';
```
Comment (klimov-paul) [inline framework/mail/BaseMailer.php]: Here change also make a little sense. Email message file is meant to be created for debugging purposes and does not affect security anyhow.

**Review by klimov-paul — CHANGES_REQUESTED** [review]

File: framework/caching/DbCache.php (hunk 1 above)
Comment (yyxx9988) [inline framework/caching/DbCache.php]: agreed.

Comment (samdark) [inline framework/caching/DbCache.php]: @yyxx9988 could you change this one back then? And the one below.

Comment (yyxx9988) [inline framework/caching/DbCache.php]: @samdark updated and changed one more file `UploadedFileTest.php`
```php
    private function generateFakeFileData()
    {
        return [
            'name' => md5(random_int()),
            'tmp_name' => md5(random_int()),
            'type' => 'image/jpeg',
            'size' => random_int(1000, 10000),
            'error' => 0,
        ];
    }
```

**Review by klimov-paul — APPROVED** [review]

**Comment by yyxx9988** [comment]
closed after #14632

**Comment by klimov-paul** [comment]
If you are unable to perform a merge - open a new PR.

**Comment by samdark** [comment]
Merged. Thanks!

**Pull request merged**